### PR TITLE
[#11153] add null-check in ConcatIterator.advance

### LINKED ISSUE #11153: Iteration over exhausted sub-iterator leads to NPE
```
val a = Iterator.empty ++ Iterator.empty
a.hasNext
val b = Iterator.empty ++ a
while (b.hasNext) b.next()

Exception in thread "main" java.lang.NullPointerException
	at scala.collection.Iterator$ConcatIterator.advance(Iterator.scala:195)
	at scala.collection.Iterator$ConcatIterator.hasNext(Iterator.scala:222)
	at Test$.delayedEndpoint$Test$1(Test.scala:19)
	at Test$delayedInit$body.apply(Test.scala:13)
	at scala.Function0.apply$mcV$sp(Function0.scala:34)
	at scala.Function0.apply$mcV$sp$(Function0.scala:34)
	at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:12)
	at scala.App.$anonfun$main$1$adapted(App.scala:76)
	at scala.collection.immutable.List.foreach(List.scala:389)
	at scala.App.main(App.scala:76)
	at scala.App.main$(App.scala:74)
	at Test$.main(Test.scala:13)
	at Test.main(Test.scala)
```
The reason is that internals of ConcatIterator [null'ed](https://github.com/scala/scala/blame/2.13.x/src/library/scala/collection/Iterator.scala#L1044) on `hasNext` method invocation, but `advance` invokes [current.hasNext](https://github.com/scala/scala/blame/2.13.x/src/library/scala/collection/Iterator.scala#L1053) without checking it for null. 

Works in 2.11.8. Doesn't work in 2.12.6/2.13-M5.

### PR DESCRIPTION
**Pull request by tIGO**
fix https://github.com/scala/bug/issues/11153

### CODE REVIEW

**Comment by SethTisue** [comment]
I wonder if we should try to sneak this into 2.12.7...

File: src/library/scala/collection/Iterator.scala [hunk 1]
```diff
@@ -1050,7 +1050,7 @@ object Iterator extends IterableFactory[Iterator] {
         tail = tail.tail
         merge()
         if (currentHasNextChecked) true
-        else if (current.hasNext) {
+        else if ((current ne null) && current.hasNext) {
```
Comment (SethTisue) [inline src/library/scala/collection/Iterator.scala:1053]: tiny nitpick: you can write `current != null`, it's more idiomatic, equally performant (same bytecode generated), and doesn't require the extra set of parens

Comment (SethTisue) [reply src/library/scala/collection/Iterator.scala:1053]: I see the file has a few other `ne null`s, though, so I guess cleaning that up is outside the scope of this PR. so never mind, I guess

Comment (tIGO) [reply src/library/scala/collection/Iterator.scala:1053]: > equally performant (same bytecode generated)

Hm, I don't know that, thanks. Is it true only for null-check?

Comment (SethTisue) [reply src/library/scala/collection/Iterator.scala:1053]: I'm not sure exactly what you mean by "it". https://www.scala-lang.org/files/archive/spec/2.12/12-the-scala-standard-library.html is the place to start if you want to dig deeper into `==`/`!=`/`eq`/`ne`

**Comment by adriaanm** [comment]
Sorry, 2.12.7 is closed now, but great to have this in for 2.13.0-RC1!

**Comment by tIGO** [comment]
Any plans for 2.12.8?

**Comment by adriaanm** [comment]
ASAP after 2.13.0, which will require our full attention for the next few months (I'll share an update this week on RC1 planning). Our schedule is at https://github.com/scala/scala/milestones

**Comment by adriaanm** [comment]
If for some reason the current candidate for 2.12.7 doesn't work out, we'll also include this fix.  Could you already back port your PR to 2.12.x (and prefix the commit title with [nomerge])

**Comment by tIGO** [comment]
Sure, https://github.com/scala/scala/pull/7227.

**Comment by SethTisue** [comment]
the backport will be included in 2.12.7.

**Pull request merged**